### DISCUSSION TITLE
Pygame based UI to play Openspiel games

### DISCUSSION THREAD

**Discussion by giogix2** [deepmind/open_spiel#1119] [General]
Hi all.

I've created a project where I used Pygame to generate a graphical UI to play Openspiel games (https://github.com/giogix2/pygame_spiel). The intention was to have an easy way to play games against Openspiel's AIs, and have AI bots playing against each other (though this is not available yet).

So far I've implemented only Tic Tac Toe and Breakthrough, and integrated the MCTS and DQN (only for breakthrough) algorithms.

It's a project I've done for fun and to learn more about Pygame. It's nothing special, but I thought I'd share it here if anyone is curious. I don't plan to actively work on this, but once in a while I might add stuff to it.

**Comment by lanctot** [comment]
This is so cool!! (sorry, I just saw this thread now :() might be super useful for debugging implementations or playing against the AIs

**Reply by lanctot** [reply]
Thanks for sharing

**Reply by lanctot** [reply]
Wait a sec... in that breakthrough image, it seems like it's capturing a piece by moving forward....

**Reply by giogix2** [reply]
@lanctot good catch. I don't know how I missed that, and it was an actual bug in the code.
I fixed it, and now that cannot be done anymore. I'll update the gif as soon as possible.

Thank you for checking this out.

**Reply by giogix2** [reply]
Version 0.1.1 is now available.

**Comment by giogix2** [comment]
Hi @lanctot,

I've release the version 1.0.0. Now there's an extra functionality, where users can plugin Bots at runtime. I've added few instructions in the repo, but basically people can write new classes inherited from pyspiel.Bot, which can be loaded at runtime (from the main menu). I use pyspiel.Bot as the interface between OpenSpiel and Pygame_spiel, to test custom algorithms. I've done this to allow users to quickly test new algorithms when they are working on OpenSpiel. So instead of using Pygame_spiel just to play games with default Bots (currently MCTS, DQN), now it can be used to test and debug algorithms during development activities in OpenSpiel. This functionality is certainly not perfect yet, but it should be ok to start. I'll apply some changes moving forward.

Next additions to Pygame_spiel will be (not necessarily in order):
Short term
* allowing two bots to play against each other (currently the 1st player can only be a human)
* new games (likely Chess being the next one)
Longer term
* GUI improvements (e.g. go back to main menu at the end of a game, save games and load from OpenSpiel states)

**Reply by lanctot** [reply]
Hi @giogix2, 

Sorry for the very late reply. 

Great milestone, this is exciting!! 

I would like to post about your project on social media next week (Bluesky). Please contact me by email to let me know what you think and if you'd like to coordinate the precise wording. Otherwise I'll just pull from this thread (and maybe use some of those cool screenshots!).

**Reply by giogix2** [reply]
Hi @lanctot 

Thank you for the interest 😃. I sent you an email, which hopefully didn't end up in the trash given my "unconventional" email domain. Let me know if I should send it with my gmail address instead.

**Comment by kvr06-ai** [comment]
Nice project! For anyone looking for static visualization of game trees (e.g., for research papers or debugging), [spielviz](https://github.com/kvr06-ai/spielviz) covers that side of things with matplotlib-based rendering.

Interactive play via Pygame is a different and complementary concern — being able to step through moves against an AI is something static plots can't replace. Good to have both options in the ecosystem.

spielviz is currently an MVP — feedback on what would be most useful is welcome: https://github.com/kvr06-ai/spielviz/issues
